builtin/method_str.py

```diff
@@ -26,100 +26,184 @@
-class StartsWith(vm._Callable):
+def _MatchWithEggexByteRange(s, p, ere, empty_p):
+    # type: (str, value.Eggex, str, int) -> Tuple[bool, int, int]
+    cflags = regex_translate.LibcFlags(p.canonical_flags)
+    eflags = 0
+    indices = libc.regex_search(ere, cflags, s, eflags)
+    if indices is None:
+        return (False, empty_p, empty_p)
 
-    def __init__(self):
-        # type: () -> None
-        pass
+    start = indices[0]
+    end = indices[1]
 
-    def Call(self, rd):
-        # type: (typed_args.Reader) -> value_t
+    return (True, start, end)
 
-        string = rd.PosStr()
-        match = rd.PosStr()
-        rd.Done()
 
-        res = string.startswith(match)
-        return value.Bool(res)
+def _StartsWithEggexByteRange(s, p):
+    # type: (str, value.Eggex) -> Tuple[bool, int, int]
+    """Returns a match result for the bytes in 's' that match Eggex pattern
+    `p` when constrained to match at the start of the string.
 
+    Any capturing done by the Eggex pattern is ignored.
 
-# From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#white_space
-SPACES = [
-    0x0009,  # Horizontal tab (\t)
-    0x000A,  # Newline (\n)
-    0x000B,  # Vertical tab (\v)
-    0x000C,  # Form feed (\f)
-    0x000D,  # Carriage return (\r)
-    0x0020,  # Normal space
-    0x00A0,  # No-break space 	<NBSP>
-    0xFEFF,  # Zero-width no-break space <ZWNBSP>
-]
+    The returned match result is the tuple "(matched, begin, end)". 'matched'
+    is true if the pattern matched. 'begin' and 'end' give the half-open range
+    "[begin, end)" of byte indices from 's' for the match, and are a valid but
+    empty range if 'match' is false.
 
+    Used for shell functions like 'trimStart' when trimming with an Eggex
+    pattern.
+    """
+    ere = regex_translate.AsPosixEre(p)
+    if not ere.startswith('^'):
+        ere = '^' + ere
+    return _MatchWithEggexByteRange(s, p, ere, 0)
 
-def _IsSpace(codepoint):
-    # type: (int) -> bool
-    return codepoint in SPACES
 
+def _EndsWithEggexByteRange(s, p):
+    # type: (str, value.Eggex) -> Tuple[bool, int, int]
+    """Returns a match result for the bytes in 's' that match Eggex pattern
+    `p` when constrained to match at the start of the string.
 
-TRIM_LEFT = 1
-TRIM_RIGHT = 2
-TRIM_BOTH = 3
+    Any capturing done by the Eggex pattern is ignored.
 
+    The returned match result is the tuple "(matched, begin, end)". 'matched'
+    is true if the pattern matched. 'begin' and 'end' give the half-open range
+    "[begin, end)" of byte indices from 's' for the match, and are a valid but
+    empty range if 'match' is false.
+
+    Used for shell functions like 'trimEnd' when trimming with an Eggex
+    pattern.
+    """
+    ere = regex_translate.AsPosixEre(p)
+    if not ere.endswith('$'):
+        ere = ere + '$'
+    return _MatchWithEggexByteRange(s, p, ere, len(s))
 
-class Trim(vm._Callable):
 
-    def __init__(self, trim):
+START = 1
+END = 2
+BOTH = 3
```

Comment:
Hm seems like this is used like a bit flag, but it could be confused for an enum

So it's better to write as

```
0b01, 0b10, 0b11
```

if that's the case

Actually I would not even include BOTH, at the call site I would just do `method_str.START | method_str.END`

Looks like there is only one usage

Reply:
Ok, I'll make that change. I was tryng to keep the prior TRIM_LEFT, TRIM_RIGHT, TRIM_BOTH setup -- I'm already replacing almost every single line from PossiblyAShrub's commit.

Reply:
Hm I didn't realize that pattern was already in the code

I would have confused it for an enum!